framework/src/csifw/CSIService.c

```diff
@@ -0,0 +1,239 @@
+	if (gDataDumpDescriptor < 0) {
+		CSIFW_LOGE("open CSI_DUMP_DATA_QUEUE fail %d", errno);
+		unlink(CSI_DUMP_DATA_QUEUE_NAME);
+		return CSIFW_ERROR;
+	}
+	result = csi_logs_dumper_init();
+	if (result != CSIFW_OK) {
+		CSIFW_LOGE("open csi_logs_dumper_init fail");
+		close(gDataDumpDescriptor);
+		unlink(CSI_DUMP_DATA_QUEUE_NAME);
+		return CSIFW_ERROR;
+	}
+	set_event_listener(csi_data_dump_listener);
+	#endif
+
+	g_ptr = ptr;
+	if (g_service_state != CSI_STATE_UNITIALIZED) {
+		CSIFW_LOGE("Already Initialized");
+		return CSIFW_ERROR;
+	}
+
+	CSIFW_RES res = CSIFW_OK;
+	g_raw_callback = raw_callback;
+	g_parsed_callback = parsed_callback;
+	csi_set_data_collect_interval(interval);
+	gParsedDataBufferLen = CSIFW_MAX_RAW_BUFF_LEN;
+	if (g_parsed_callback) {
+		g_parsed_buffptr = (float *)malloc(sizeof(float) * gParsedDataBufferLen);
+		if (!g_parsed_buffptr) {
+			CSIFW_LOGE("Failed to allocate memory for parsed data buffer, size: %zu", sizeof(float) * gParsedDataBufferLen);
+			return CSIFW_ERROR;
+		}
+	}
+	g_service_state = CSI_STATE_INITIALIZED;
+
+	CSIFW_RES ret;
+	ret = csi_packet_receiver_init(CSIRawDataListener);
+
+	if (ret != CSIFW_OK) {
+		CSIFW_LOGE("csi_packet_receiver_init failed");
+		return CSIFW_ERROR;
+	}
+	CSIFW_LOGD("CSI Packet receiver init done");
+
+	return res;
+}
+
+CSIFW_RES csi_set_data_collect_interval(unsigned interval)
+{
+	return ping_generator_change_interval(interval);
+}
+
+CSIFW_RES csi_service_start()
+{	
+	if (g_service_state == CSI_COLLECT_STATE_STARTED) {
+		CSIFW_LOGE(" Already running");
+		return CSIFW_ERROR;
+	}
+	csi_packet_receiver_start_collect(&g_csifw_config);
+	ping_generator_start();
+	g_service_state = CSI_COLLECT_STATE_STARTED;
+	return CSIFW_OK;
+}
+
+CSIFW_RES csi_service_stop(CSIFW_REASON reason)
+{
+	if (g_service_state != CSI_COLLECT_STATE_STARTED) {
+		CSIFW_LOGE("Already stopped");
+		g_service_state = CSI_COLLECT_STATE_STOPPED;
+		return CSIFW_ERROR;
+	}
+	g_service_state = CSI_COLLECT_STATE_STOPPED;
+	if (reason == CSIFW_WIFI_DISCONNECTED)
+	{
+		CSIFW_LOGI("[CSIFW_REASON] CSIFW_WIFI_DISCONNECTED %d", reason);
+	}
+	ping_generator_stop(reason);
+	csi_packet_receiver_stop_collect(reason, &g_csifw_config);
+	return CSIFW_OK;
+}
+
+CSIFW_RES csi_service_set_config(csi_action_param_t config)
+{
+	g_csifw_config = config; 
+	return CSIFW_OK;
+}
+
+CSIFW_RES csi_get_ap_mac_addr(csifw_mac_info *mac_info) {
+	return csi_packet_receiver_get_mac_addr(mac_info);
+}
+
+CSIFW_RES csi_service_deinit()
+{
+	CSIFW_LOGD("csi_service_deinit");
+	if (g_service_state != CSI_COLLECT_STATE_STOPPED) {
+		csi_service_stop(CSIFW_NORMAL);
+	}
+	g_service_state = CSI_STATE_UNITIALIZED;
+	csi_packet_receiver_deinit();
+	#ifdef CONFIG_CSI_DATA_DUMP_OVER_NETWORK
+	csi_logs_dumper_deinit();
+	close(gDataDumpDescriptor);
+	unlink(CSI_DUMP_DATA_QUEUE_NAME);
+	#endif
+	if (g_parsed_buffptr) {
+		free(g_parsed_buffptr);
+		g_parsed_buffptr = NULL;
+		CSIFW_LOGD("csi_service_deinit");
```

Comment:
Is it ok to print this when g_parsed_buffer is not null?
If it's a NULL, then this message will not be printed. But the function returns OK.